flink-end-to-end-tests/flink-rest-api-test/src/main/java/org/apache/flink/runtime/rest/tests/RestApiTestSuite.java

```diff
@@ -0,0 +1,363 @@
+
+	public static void main(String[] args) throws Exception {
+
+		ParameterTool params = ParameterTool.fromArgs(args);
+		final String host = params.get("host", "localhost");
+		final int port = params.getInt("port", 8081);
+		final HttpTestClient httpClient = new HttpTestClient(host, port);
+
+		// Validate Flink cluster is running
+		JobIdsWithStatusOverview jobOverview = getJobOverview(httpClient);
+
+		// Get necessary parameters for testing GET endpoints
+		Map<String, String> parameterMap = getParameterMaps(httpClient, jobOverview);
+
+		// Get list of endpoints
+		List<MessageHeaders> specs = new E2ETestDispatcherRestEndpoint().getSpecs();
+		specs.forEach(spec -> testMonitoringEndpointSpecs(httpClient, spec, parameterMap));
+
+		if (testFailureCount != 0) {
+			throw new RuntimeException("There are test failures. Success: " + testSuccessCount +
+				" Failures: " + testFailureCount + " Skipped: " + testSkipCount);
+		}
+	}
+
+	@SuppressWarnings("ConstantConditions")
+	private static Map<String, String> getParameterMaps(HttpTestClient httpClient,
+		JobIdsWithStatusOverview jobOverview) throws InterruptedException, IOException, TimeoutException {
+		// Get necessary parameters used for all REST API testings.
+		final Map<String, String> parameterMap = new HashMap<>();
+		Preconditions.checkState(jobOverview.getJobsWithStatus().stream()
+			.filter(jobIdWithStatus -> jobIdWithStatus.getJobStatus() == JobStatus.RUNNING)
+			.count() >= 1, "Cannot found active running jobs, discontinuing test!");
+		String jobId = jobOverview.getJobsWithStatus().stream()
+			.filter(jobIdWithStatus -> jobIdWithStatus.getJobStatus() == JobStatus.RUNNING)
+			.findFirst().get().getJobId().toString();
+		parameterMap.put(":jobid", jobId);
+
+		JobDetailsInfo jobDetailsInfo = getJobDetailInfo(httpClient, jobId);
+		String vertexId = jobDetailsInfo.getJobVertexInfos().stream()
+			.findFirst().get().getJobVertexID().toString();
+		parameterMap.put(":vertexid", vertexId);
+		parameterMap.put(":checkpointid", "1"); // test first checkpoint
+		parameterMap.put(":subtaskindex", "0"); // test first subtask
+
+		TaskManagersInfo taskManagersInfo = getTaskManagers(httpClient);
+		String taskMgrId = taskManagersInfo.getTaskManagerInfos().stream().findFirst().get().getResourceId().toString();
+		parameterMap.put(":taskmanagerid", taskMgrId);
+		parameterMap.put(":triggerid", "");
+
+		return parameterMap;
+	}
+
+	private static JobIdsWithStatusOverview getJobOverview(HttpTestClient httpClient)
+		throws TimeoutException, InterruptedException, IOException {
+		httpClient.sendGetRequest("/jobs", TEST_TIMEOUT);
+		HttpTestClient.SimpleHttpResponse resp = httpClient.getNextResponse();
+		Preconditions.checkState(resp.getStatus().code() == 200,
+			"Cannot fetch Flink cluster status!");
+		return MAPPER.readValue(resp.getContent(), JobIdsWithStatusOverview.class);
+	}
+
+	private static JobDetailsInfo getJobDetailInfo(HttpTestClient httpClient, String jobId)
```

Comment:
these methods can be made less brittle my using the corresponding `MessageHeader` objects.

```
JobDetailsHeaders headers = JobDetailsHeaders.getInstance();
httpClient.sendGetRequest(headers.getTargetRestEndpointURL().replace(JobIDPathParameter.KEY, jobId), TEST_TIMEOUT);
HttpTestClient.SimpleHttpResponse resp = httpClient.getNextResponse();
Preconditions.checkState(resp.getStatus() == headers.getResponseStatusCode(),
	"Cannot fetch job detail information for job " + jobId);
return MAPPER.readValue(resp.getContent(), headers.getResponseClass());
```